### ISSUE TITLE
HTTP::Tiny is broken on older perls

### ISSUE THREAD

**Issue by dur-randir** [chansen/p5-http-tiny#109]
HTTP::Tiny works only on perls >= 5.20. On 5.18 and earlier, you get the following:

```
t/000_load.t ........... 1/1
#   Failed test 'require HTTP::Tiny;'
#   at t/000_load.t line 8.
#     Tried to require 'HTTP::Tiny'.
#     Error:  "EINTR" is not exported by the Errno module
#  "EPIPE" is not exported by the Errno module
```

This breaks CPAN client update on all those perls, since it depends on HTTP::Tiny when it's available. Here's an example of perl -V:

```
Summary of my perl5 (revision 5 version 18 subversion 0) configuration:

  Platform:
    osname=linux, osvers=4.9.0-3-amd64, archname=x86_64-linux
    uname='linux dev.crazypanda.ru 4.9.0-3-amd64 #1 smp debian 4.9.30-2+deb9u5 (2017-09-19) x86_64 gnulinux '
    config_args='-de -Dprefix=/home/s.aleinikov/perlbrew/perls/perl-5.18.0 -Dman1dir=none -Dman3dir=none -Aeval:scriptdir=/home/s.aleinikov/perlbrew/perls/perl-5.18.0/bin'
    hint=recommended, useposix=true, d_sigaction=define
    useithreads=undef, usemultiplicity=undef
    useperlio=define, d_sfio=undef, uselargefiles=define, usesocks=undef
    use64bitint=define, use64bitall=define, uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='cc', ccflags ='-fno-strict-aliasing -pipe -fstack-protector -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64',
    optimize='-O2',
    cppflags='-fno-strict-aliasing -pipe -fstack-protector -I/usr/local/include'
    ccversion='', gccversion='6.3.0 20170516', gccosandvers=''
    intsize=4, longsize=8, ptrsize=8, doublesize=8, byteorder=12345678
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=16
    ivtype='long', ivsize=8, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=8, prototype=define
  Linker and Libraries:
    ld='cc', ldflags =' -fstack-protector -L/usr/local/lib'
    libpth=/usr/local/lib /lib/x86_64-linux-gnu /lib/../lib /usr/lib/x86_64-linux-gnu /usr/lib/../lib /lib /usr/lib /lib64 /usr/lib64
    libs=-lnsl -ldl -lm -lcrypt -lutil -lc
    perllibs=-lnsl -ldl -lm -lcrypt -lutil -lc
    libc=libc-2.24.so, so=so, useshrplib=false, libperl=libperl.a
    gnulibc_version='2.24'
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-Wl,-E'
    cccdlflags='-fPIC', lddlflags='-shared -O2 -L/usr/local/lib -fstack-protector'


Characteristics of this binary (from libperl):
  Compile-time options: HAS_TIMES PERLIO_LAYERS PERL_DONT_CREATE_GVSV
                        PERL_HASH_FUNC_ONE_AT_A_TIME_HARD PERL_MALLOC_WRAP
                        PERL_PRESERVE_IVUV PERL_SAWAMPERSAND USE_64_BIT_ALL
                        USE_64_BIT_INT USE_LARGE_FILES USE_LOCALE
                        USE_LOCALE_COLLATE USE_LOCALE_CTYPE
                        USE_LOCALE_NUMERIC USE_PERLIO USE_PERL_ATOF
  Built under linux
  Compiled at Jan 12 2018 19:43:12
  %ENV:
    PERLBREW_BASHRC_VERSION="0.71"
    PERLBREW_HOME="/home/s.aleinikov/.perlbrew"
    PERLBREW_MANPATH="/home/s.aleinikov/perlbrew/perls/perl-5.18.0/man"
    PERLBREW_PATH="/home/s.aleinikov/perlbrew/bin:/home/s.aleinikov/perlbrew/perls/perl-5.18.0/bin"
    PERLBREW_PERL="perl-5.18.0"
    PERLBREW_ROOT="/home/s.aleinikov/perlbrew"
    PERLBREW_VERSION="0.71"
  @INC:
    /home/s.aleinikov/perlbrew/perls/perl-5.18.0/lib/site_perl/5.18.0/x86_64-linux
    /home/s.aleinikov/perlbrew/perls/perl-5.18.0/lib/site_perl/5.18.0
    /home/s.aleinikov/perlbrew/perls/perl-5.18.0/lib/5.18.0/x86_64-linux
    /home/s.aleinikov/perlbrew/perls/perl-5.18.0/lib/5.18.0
```

**Comment by chansen** [comment]
GCC 5.0 changed the output from the C preprocessor: [Bug #123784 for perl5: Tests fail with GCC 5.0 because Errno cannot obtain errno constant](https://rt.perl.org/Public/Bug/Display.html?id=123784) (Fixed in Perl v5.22, patch: <https://github.com/Perl/perl5/commit/816b056>).

**Comment by chansen** [comment]
@dur-randir I'm sorry for incorrectly assuming that you forced installed or skipped testing when installing Perl!

I'm smacked by the fact that the Perl test suite didn't catch that `Errno.pm` failed to parse/obtain the `errno.h` error codes.

This brings us to the question, how do we obtain the `errno.h` error codes in a portable way?

**Comment by xdg** [comment]
HTTP::Tiny builds and tests perfectly well on older Perls -- as long as those Perls are correctly built.  See the Travis output here: https://travis-ci.org/chansen/p5-http-tiny

There have been similar compiler-related issues over the years.  If your provider of Perl does not correctly patch Perl to account for these differences and you have a broken Perl, that's not our responsibility.   (See, for example https://github.com/gugod/App-perlbrew/issues/448)

Please take this problem upstream to perlbrew.

**Comment by dur-randir** [comment]
The same thing is broken with the following perl:
```
Summary of my perl5 (revision 5 version 18 subversion 0) configuration:

  Platform:
    osname=linux, osvers=4.9.0-3-amd64, archname=x86_64-linux
    uname='linux dev.crazypanda.ru 4.9.0-3-amd64 #1 smp debian 4.9.30-2+deb9u5 (2017-09-19) x86_64 gnulinux '
    config_args='-Dman1dir=none -Dman3dir=none'
    hint=recommended, useposix=true, d_sigaction=define
    useithreads=undef, usemultiplicity=undef
    useperlio=define, d_sfio=undef, uselargefiles=define, usesocks=undef
    use64bitint=define, use64bitall=define, uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='cc', ccflags ='-fno-strict-aliasing -pipe -fstack-protector -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64',
    optimize='-O2',
    cppflags='-fno-strict-aliasing -pipe -fstack-protector -I/usr/local/include'
    ccversion='', gccversion='6.3.0 20170516', gccosandvers=''
    intsize=4, longsize=8, ptrsize=8, doublesize=8, byteorder=12345678
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=16
    ivtype='long', ivsize=8, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=8, prototype=define
  Linker and Libraries:
    ld='cc', ldflags =' -fstack-protector -L/usr/local/lib'
    libpth=/usr/local/lib /lib/x86_64-linux-gnu /lib/../lib /usr/lib/x86_64-linux-gnu /usr/lib/../lib /lib /usr/lib /lib64 /usr/lib64
    libs=-lnsl -ldl -lm -lcrypt -lutil -lc
    perllibs=-lnsl -ldl -lm -lcrypt -lutil -lc
    libc=, so=so, useshrplib=false, libperl=libperl.a
    gnulibc_version='2.24'
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-Wl,-E'
    cccdlflags='-fPIC', lddlflags='-shared -O2 -L/usr/local/lib -fstack-protector'


Characteristics of this binary (from libperl):
  Compile-time options: HAS_TIMES PERLIO_LAYERS PERL_DONT_CREATE_GVSV
                        PERL_HASH_FUNC_ONE_AT_A_TIME_HARD PERL_MALLOC_WRAP
                        PERL_PRESERVE_IVUV PERL_SAWAMPERSAND USE_64_BIT_ALL
                        USE_64_BIT_INT USE_LARGE_FILES USE_LOCALE
                        USE_LOCALE_COLLATE USE_LOCALE_CTYPE
                        USE_LOCALE_NUMERIC USE_PERLIO USE_PERL_ATOF
  Built under linux
  Compiled at Mar  8 2018 19:54:15
  %ENV:
    PERLBREW_BASHRC_VERSION="0.71"
    PERLBREW_HOME="/home/s.aleinikov/.perlbrew"
    PERLBREW_MANPATH="/home/s.aleinikov/perlbrew/perls/perl-5.18.0/man"
    PERLBREW_PATH="/home/s.aleinikov/perlbrew/bin:/home/s.aleinikov/perlbrew/perls/perl-5.18.0/bin"
    PERLBREW_PERL="perl-5.18.0"
    PERLBREW_ROOT="/home/s.aleinikov/perlbrew"
    PERLBREW_VERSION="0.71"
  @INC:
    /tmp/eiH8ohY/lib/perl5/site_perl/5.18.0/x86_64-linux
    /tmp/eiH8ohY/lib/perl5/site_perl/5.18.0
    /tmp/eiH8ohY/lib/perl5/5.18.0/x86_64-linux
    /tmp/eiH8ohY/lib/perl5/5.18.0
```
Which is build from officially released 5.18.0 tarball. So, who's upstream here? P5P? They've fixed this in 5.20.0. Furthermore,  Errno docs explicitly says:
```
Importing a particular constant may not be very portable, because the import will fail on platforms that do not have that constant. A more portable way to set $! to a valid value is to use:

if (exists &Errno::EFOO) {    $! = &Errno::EFOO;}
```

So, you're using unportable interface and say that it's someone else's fault? Hope https://github.com/chansen/p5-http-tiny/pull/110 this pull request helps.

**Comment by xdg** [comment]
What *would* be a helpful patch to the Makefile.PL would be to detect a broken Errno and error out with a useful message.   That helps people with broken perl installations know what happened while not restricting usage of HTTP::Tiny for people with working perls.
